File: Python/voting_system.py
import random
def get_stored_candidate_details(candidate_details):
	if type(candidate_details) != list:
		raise TypeError("candidate_details must be a List")
	for item in candidate_details:
		if not item.isalnum():
			raise ValueError("card_number must be all numeric strings")

		
	stored_candidate_details = {}
	stored_candidate_details['Name'] = candidate_details[0]
	stored_candidate_details['Party'] = candidate_details[1]
	stored_candidate_details['VIN'] = candidate_details[2]
	return stored_candidate_details
	
def get_state_zip_code(voter_details):
	states_and_zip_code = {
	"ABIA": 440001, "ADAMAWA": 640001, "AKWA IBOM": 520001, "ANAMBRA": 420001, "BAUCHI": 740001,
	"BAYELSA": 561001, "BENUE": 970001, "BORNO": 600001, "CROSS RIVER": 540001, "DELTA": 320001,
	"EBONYI": 840001, "EDO": 300001, "EKITI": 360001, "ENUGU": 400001, "GOMBE": 760001, "IMO": 460001,
	"JIGAWA": 720001, "KADUNA": 700001, "KANO": 800001, "KASTINA": 820001, "KEBBI": 860001, "KOGI": 260001,
	"KWARA": 240001, "LAGOS": 100001, "NASARAWA": 962001, "NIGER": 920001, "OGUN": 110001, "ONDO": 340001,
	"OSUN": 230001, "OYO": 200001, "PLATEAU": 930001, "RIVERS": 500001, "SOKOTO": 840001, "TARABA": 660001,
	"YOBE": 620001, "FCT": 900001
	}
	state = voter_details[2]
	
	return states_and_zip_code[state]

def get_vin(voter_details):
	zip_code = str(get_state_zip_code(voter_details))
	reg_num = str(random.randrange(0, 1000))
	lga_abbreviation = [voter_details[2][0:3], voter_details[3][0:3]]
	vin = lga_abbreviation[0] + zip_code + reg_num + str(lga_abbreviation[1])
	return vin

def get_voters_list(voter_details):
	stored_voters_details = {}
	stored_voters_details['Name'] = voter_details[0]
	stored_voters_details['DOB'] = voter_details[1]
	stored_voters_details['State'] = voter_details[2]
	stored_voters_details['LGA'] = voter_details[3]
	stored_voters_details['Contact'] = voter_details[4]
	stored_voters_details['VIN'] = voter_details[5]
	return stored_voters_details
	
def is_valid(user_response, voters_list, cast_voters):
	for number in range(0, len(voters_list)):
		for key in voters_list[number]:
			if key == "VIN":
				if voters_list[number][key] == user_response:
					if user_response not in cast_voters:
						return "valid"
					else:
						return "not allowed"
				
	return "invalid"

def get_registered_parties(candidates_list):
	return [candidate["Party"] for candidate in candidates_list]


def get_count_of_votes(list_of_casted_votes):
	counts = {}
	for casted_votes in list_of_casted_votes:
		counts[casted_votes] = counts.get(casted_votes, 0) + 1
	return(counts)


def get_election_result(vote_count, registered_parties):
	result = []
	for index_str, count in vote_count.items():
		index = int(index_str) - 1
		party = registered_parties[index]
		result.append(f"{party} scored {count}")
	return result	

def display_final_result(result):
	largest = 0
	winner = ""
	result_sentence = ""
	for election_result in result:
		party = election_result.split("scored")[0].strip()
		score = int(election_result.split("scored")[1].strip())
		if score > largest:
			largest = score
			winner = party
			result_sentence = election_result
	return f"The winner is {winner} because {result_sentence}"
			
	

File: JAVA/VotingSystem.java
import java.util.ArrayList;
import java.util.Scanner;
import java.util.HashMap;
import java.util.Map;

public class VotingSystem{

public static ArrayList<ArrayList<String>> candidatesList = new ArrayList<>();
public static ArrayList<ArrayList<String>> votersList = new ArrayList<>();
public static void main(String[] args){
ArrayList<String> candidateDetails = new ArrayList<String>();
ArrayList<String> voterDetails = new ArrayList<String>();

Scanner input = new Scanner(System.in);
boolean list0fOptions = true;
while (list0fOptions){
	String menu = """
		1. Candidate Registration
		2. Voter Registration
		3. Voting Process
		4. Vote Counting
		5. Election Result
		0. Exit
		""";
	System.out.println(menu); 
	System.out.println("Select an option: ");
	int userResponse = input.nextInt();
	switch(userResponse){
		case 1:
			displayWelcomeNote(userResponse)	;	
			System.out.println("Enter your name: ");
			String candidateName = input.next();
			boolean isAlpha = isAllLetters(candidateName);
			while (isAlpha == false){
				System.out.print("Invalid Name. Please ");
				System.out.println("Enter your name: ");
				candidateName = input.next();
			}
			
			System.out.println("Enter your Party abreviation: ");
			String candidateParty = input.next();
			candididateParty = candididateParty.toUpperCase();
			isAlpha = isAllLetters(candidateParty);
			while (isAlpha == false){
				System.out.print("Invalid Party Abbreviation. Please ");
				System.out.println("Enter your Party abreviation: ");
				candidateParty = input.next();
			}
			
			System.out.println("Enter your voter's identification number: ");
			String candidateVin = input.next();
			boolean alphaNumeric = isAlphaNumeric(candidateVin);
			while (alphaNumeric == false){				
				System.out.print("Invalid Voter Identification number. Please ");
				System.out.println("Enter your voter's identification number: ");
				candidateVin = input.next();
			}

			
		
			candidateDetails.add(candidateName);
			candidateDetails.add(candidateParty);
			candidateDetails.add(candidateVin);
			
			candidatesList.add(getStoredCandidateDetails(candidateDetails));
			System.out.println(candidateDetails);
			System.out.println(candidatesList);
			
			break;
		case 2:
			displayWelcomeNote(userResponse)	;
			System.out.println("Enter your name:");
			String voterName = input.next();
			isAlpha = isAllLetters(voterName);
			while (isAlpha == false){
				System.out.println("Enter your name:");
				voterName = input.next();
			}
			System.out.println("Enter your DOB: ");
			String voterDateOfBirth = input.next();
			
			System.out.println("Enter your state of origin: ");
			String voterStateOfOrigin = input.next();
			voterStateOfOrigin = voterStateOfOrigin.toUpperCase();
			isAlpha = isAllLetters(voterStateOfOrigin);
			while (isAlpha == false){
				System.out.println("Enter your state of origin: ");
				voterStateOfOrigin = input.next();
			}

			System.out.println("Enter your LGA: ");
			String voterLGA = input.next();
			isAlpha = isAllLetters(voterLGA);
			while (isAlpha == false){
				System.out.println("Enter your LGA: ");
				voterLGA = input.next();
			}

			System.out.println("Enter your Contact: ");
			String voterContact = input.next();
			boolean isDigit = isAllDigits(voterContact);
			while (isDigit == false){
				System.out.println("Enter your Contact: ");
				voterContact = input.next();
			}
			
			voterDetails.add(voterName);
			voterDetails.add(voterDateOfBirth);
			voterDetails.add(voterStateOfOrigin);
			voterDetails.add(voterLGA);
			voterDetails.add(voterContact);
			
			String voterVIN = getVIN(voterDetails);
			System.out.println("This is your voters identification number " + voterVIN);
			
			voterDetails.add(voterVIN);
			votersList.add(getVotersList(voterDetails));
			System.out.println(votersList);

			
	}
}

		
}

public static ArrayList<String> getStoredCandidateDetails(ArrayList<String> candidateDetails){
	ArrayList<String> storedCandidateDetails = new ArrayList<String>();
	storedCandidateDetails.add(candidateDetails.get(0));
	storedCandidateDetails.add(candidateDetails.get(1));
	storedCandidateDetails.add(candidateDetails.get(2));
	return storedCandidateDetails;
}

public static String getStateZipCode(ArrayList<String> voterDetails){
	Map<String, String> stateZipCode = new HashMap<>();
	
	stateZipCode.put("ABIA", "440001");  stateZipCode.put("ADAMAWA", "640001");
	stateZipCode.put("AKWA IBOM", "520001");  stateZipCode.put("ANAMBRA", "420001");
	stateZipCode.put("BAUCHI", "740001");  stateZipCode.put("BAYELSA", "561001");
	stateZipCode.put("BENUE", "970001");  stateZipCode.put("BORNO", "600001");
	stateZipCode.put("CROSS RIVER", "540001");  stateZipCode.put("DELTA", "320001");
	stateZipCode.put("EBONYI", "840001");  stateZipCode.put("EDO", "300001");
	stateZipCode.put("EKITI", "360001");  stateZipCode.put("ENUGU", "400001");
	stateZipCode.put("GOMBE", "760001");  stateZipCode.put("IMO", "460001");
	stateZipCode.put("JIGAWA", "720001");  stateZipCode.put("KADUNA", "700001");
	stateZipCode.put("KANO", "800001");  stateZipCode.put("KASTINA", "820001");
	stateZipCode.put("KEBBI", "860001");  stateZipCode.put("KOGI", "260001");
	stateZipCode.put("KWARA", "240001");  stateZipCode.put("LAGOS", "100001");
	stateZipCode.put("NASARAWA", "962001,");  stateZipCode.put("NIGER", "920001");
	stateZipCode.put("OGUN", "110001");  stateZipCode.put("ONDO", "340001");
	stateZipCode.put("OSUN", "230001");  stateZipCode.put("OYO", "200001");
	stateZipCode.put("PLATEAU", "930001");  stateZipCode.put("RIVERS", "500001");
	stateZipCode.put("SOKOTO", "840001");  stateZipCode.put("TARABA", "660001");
	stateZipCode.put("YOBE", "620001");  stateZipCode.put("FCT", "900001");

	String state = voterDetails.get(2);
	return stateZipCode.get(state);
}

public static String getVIN(ArrayList<String> voterDetails){
	String zipCode = getStateZipCode(voterDetails);
	String regNum = "" + (int)(Math.random() * 1000);
	String lgaAbbreviation = voterDetails.get(2);
	lgaAbbreviation = lgaAbbreviation.substring(0, 3);
	lgaAbbreviation += zipCode;
	lgaAbbreviation += regNum;
	lgaAbbreviation += (voterDetails.get(3)).substring(0, 3);
	return lgaAbbreviation;
}

public static ArrayList<String> getVotersList(ArrayList<String> voterDetails){
	ArrayList<String> storedVotersDetails = new ArrayList<String>();
	storedVotersDetails.add(voterDetails.get(0));
	storedVotersDetails.add(voterDetails.get(1));
	storedVotersDetails.add(voterDetails.get(2));
	storedVotersDetails.add(voterDetails.get(3));
	storedVotersDetails.add(voterDetails.get(4));
	storedVotersDetails.add(voterDetails.get(5));
	return storedVotersDetails;
}


public static boolean isAllLetters(String word){
	boolean isAlpha = true;
	for(char c: word.toCharArray()){
		if(!Character.isLetter(c)){
			isAlpha = false;
			break;
		}
	}
	return isAlpha;
}
	
public static boolean isAlphaNumeric(String wordNum){
	for(char c : wordNum.toCharArray()){
		if(!Character.isLetterOrDigit(c)){
			return false;
		}
	}
	return true;
	
}

public static boolean isAllDigits(String number){
	for(char c: number.toCharArray()){
		if (!Character.isDigit(c)){
			return false;
		}
	}
	return true;
}

public static void displayWelcomeNote(int number){
	if (number == 1){
		System.out.println("""
		===============================================
		Welcome to the registration site for candidates
		Fill the form below:
		================================================
		""");
	}
	else if (number == 2){
		System.out.println("""
		===============================================
		Welcome to the registration site for voters
		Fill the form below:
		================================================
		""");
	}

}



	
	


}
	